okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -439,6 +464,69 @@ public void close() {}
     return null;
   }
 
+  private Socket createHttpProxySocket(InetSocketAddress address, InetSocketAddress proxyAddress,
+      String proxyUsername, String proxyPassword) {
+    Status internalStatus = Status.INTERNAL.withDescription("Failed trying to connect with proxy");
+    try {
+      Socket sock = new Socket(proxyAddress.getAddress(), proxyAddress.getPort());
+      sock.setTcpNoDelay(true);
+
+      BufferedSource source = Okio.buffer(Okio.source(sock));
+      BufferedSink sink = Okio.buffer(Okio.sink(sock));
+
+      // Prepare headers and request method line
+      Request proxyRequest = createHttpProxyRequest(address, proxyUsername, proxyPassword);
+      HttpUrl url = proxyRequest.httpUrl();
+      String requestLine = String.format("CONNECT %s:%d HTTP/1.1", url.host(), url.port());
+
+      // Write request to socket
+      sink.writeUtf8(requestLine).writeUtf8("\r\n");
+      for (int i = 0, size = proxyRequest.headers().size(); i < size; i++) {
+        sink.writeUtf8(proxyRequest.headers().name(i))
+            .writeUtf8(": ")
+            .writeUtf8(proxyRequest.headers().value(i))
+            .writeUtf8("\r\n");
+      }
+      sink.writeUtf8("\r\n");
+      // Flush buffer (flushes socket and sends request)
+      sink.flush();
+
+      // Read response, parse headers, check if 200 OK was returned
+      StatusLine statusLine = StatusLine.parse(source.readUtf8LineStrict());
+      if (statusLine.code != 200) {
+        throw new IOException("Response returned from proxy was not 200 OK");
+      }
+      return sock;
+
+    } catch (SocketException e) {
+      // NB(lukaszx0) We'd use mulit-catch, but it's not supported at this language level.
+      throw internalStatus.withCause(e).asRuntimeException();
+    } catch (IOException e) {
+      throw internalStatus.withCause(e).asRuntimeException();
+    }
+  }
+
+  private Request createHttpProxyRequest(InetSocketAddress address, String proxyUsername,
+      String proxyPassword) {
+    HttpUrl tunnelUrl = new HttpUrl.Builder()
+        .scheme("https")
+        .host(address.getHostName())
```

Comment:
HostName is not appropriate. "Note: This method may trigger a name service reverse lookup if the address was created with a literal IP address." We want deterministic behavior.

Given, https://github.com/grpc/grpc/pull/8248/files#diff-194d5694ec421f672a909ccc72d922e4R41 , it's not clear to me what value should be used when, since "using a LB" is a vague notion in the code. We have an IP address and an authority, where the authority may have been overriden by the user. It seems we can only use an IP reliably. Maybe some other code needs to be reorganized in our implementation? @jboeuf @markdroth